File: common/src/main/java/com/ssblur/scriptor/color/DyeColorableBlock.java
package com.ssblur.scriptor.color;

import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import com.ssblur.scriptor.color.interfaces.ColorableBlock;
import com.ssblur.scriptor.registry.colorable.ColorableBlockRegistry;
import net.minecraft.core.BlockPos;
import net.minecraft.core.component.DataComponents;
import net.minecraft.world.item.DyeColor;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.component.DyedItemColor;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.entity.BlockEntity;

import java.util.Random;

public class DyeColorableBlock implements ColorableBlock {
  BiMap<DyeColor, Block> blocks = HashBiMap.create();
  Random random = new Random();

  @Override
  public void setColor(int color, Level level, BlockPos blockPos) {
    BlockEntity blockEntity = level.getBlockEntity(blockPos);
    level.setBlock(
      blockPos,
      blocks.get(
        CustomColors.getDyeColor(color, level.getGameTime())
      ).withPropertiesOf(level.getBlockState(blockPos)),
      2
    );
    if(blockEntity != null) {
      level.removeBlockEntity(blockPos);
      level.setBlockEntity(blockEntity);
    }
  }

  @Override
  public ItemStack setColor(int color, ItemStack itemStack) {
    var block = blocks.get(CustomColors.getDyeColor(color, Math.abs(random.nextInt())));
    var dyedItemColor = itemStack.get(DataComponents.DYED_COLOR);
    boolean tooltip = true;
    if(dyedItemColor != null)
      tooltip = dyedItemColor.showInTooltip();
    int count = itemStack.getCount();

    var itemOut = new ItemStack(block);
    itemOut.setCount(count);
    itemOut.set(DataComponents.DYED_COLOR, new DyedItemColor(color, tooltip));
    return itemOut;
  }

  public void add(Block block, DyeColor dyeColor) {
    blocks.put(dyeColor, block);
  }

  public void register() {
    for(var entry: blocks.values())
      ColorableBlockRegistry.register(entry, this);
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/color/DyeColorableBlock.kt
package com.ssblur.scriptor.color

import com.google.common.collect.BiMap
import com.google.common.collect.HashBiMap
import com.ssblur.scriptor.color.interfaces.ColorableBlock
import com.ssblur.scriptor.registry.colorable.ColorableBlockRegistry.register
import net.minecraft.core.BlockPos
import net.minecraft.core.component.DataComponents
import net.minecraft.world.item.DyeColor
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.component.DyedItemColor
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.Block
import java.util.*
import kotlin.math.abs

class DyeColorableBlock: ColorableBlock {
  var blocks: BiMap<DyeColor, Block> = HashBiMap.create()
  var random: Random = Random()

  override fun setColor(color: Int, level: Level?, blockPos: BlockPos) {
    if (level == null) return
    val blockEntity = level.getBlockEntity(blockPos)
    level.setBlock(
      blockPos,
      blocks[CustomColors.getDyeColor(color, level.gameTime.toFloat())]!!.withPropertiesOf(
        level.getBlockState(
          blockPos
        )
      ),
      2
    )
    if (blockEntity != null) {
      level.removeBlockEntity(blockPos)
      level.setBlockEntity(blockEntity)
    }
  }

  override fun setColor(color: Int, itemStack: ItemStack?): ItemStack {
    val block = blocks[CustomColors.getDyeColor(
      color, abs(random.nextInt().toDouble())
        .toFloat()
    )]!!
    val dyedItemColor = itemStack?.get(DataComponents.DYED_COLOR)
    var tooltip = true
    if (dyedItemColor != null) tooltip = dyedItemColor.showInTooltip()
    val count = itemStack?.count ?: 1

    val itemOut = ItemStack(block)
    itemOut.count = count
    itemOut.set(DataComponents.DYED_COLOR, DyedItemColor(color, tooltip))
    return itemOut
  }

  fun add(block: Block, dyeColor: DyeColor) {
    blocks[dyeColor] = block
  }

  fun register() {
    for (entry in blocks.values) register(entry!!, this)
  }
}
